platform/external-system-api/src/com/intellij/openapi/externalSystem/model/DataNode.java

```diff
@@ -279,9 +279,13 @@ private void readObject(ObjectInputStream in)
 
   @Override
   public int hashCode() {
-    int result = myChildren.hashCode();
-    result = 31 * result + myKey.hashCode();
+    int result = myKey.hashCode();
     result = 31 * result + getData().hashCode();
+    // We can't use myChildren.hashCode() because it's exponential.
```

Comment:
Exponential of a height of a tree. Or just O(amount of nodes in the sub-tree) as you mentioned. It is too much. I saw IJ was recalculating hashCode over and over again in ToolWindowModuleService#processData and just use ExternalSystemApiUtil#groupBy a lot.

Contract of hashCode method is that it should return the same value for equal objects but if it returns the same value it doesn't mean that the objects are equal. Also I don't think it will increase amount of collisions significantly but it affected performance in my testing much.

Can change the comment tho.